### PR TITLE
APIMF-2783 and APIMF-2785: add model in memory and payload validation docs 

### PR DESCRIPTION
**Pull request by arielmirra**
Add model in memory documentation and code examples

### CODE REVIEW

File: docs/other/model-in-memory.mdx [hunk 1]
```diff
@@ -0,0 +1,55 @@
+---
+id: model_in_memory
+title: Creating AMF model in memory
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:3]: Creating an AMF model in memory

File: docs/other/model-in-memory.mdx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+---
+:::info
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:5]: Is this necessary?

File: docs/other/model-in-memory.mdx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+The functionality expressed in this documentation requires familiarity with the
+[amf model](https://github.com/aml-org/amf/blob/develop/documentation/model.md).
+:::
+
+AMF is most commonly used to parse an already existing API,
+but it can also be used to create an API in a fluent and descriptive way.
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:11]: AMF allows you to parse existing APIs, but you can also use it to model fluent, descriptive APIs.

OR

AMF allows you to parse existing APIs, but you can also use it to model fluent, descriptive APIs in memory.

Is the "in memory" part important here or in the following sentence?

File: docs/other/model-in-memory.mdx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+Some use cases for creating a WebApi model in memory include:
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:13]: When you create an API [or Web API???] model in memory, you can use AMF to:

- Edit an already parsed API
- Power a visual API designer
- Create programmatic API tests

File: docs/other/model-in-memory.mdx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+- To edit an already parsed API with AMF
+- When using AMF to power a visual API designer
+- To create programmatic API tests
+
+
+### Creating a WebApi with AMF programmatically
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:20]: ## Create a Web API model with AMF

I'm not sure how we're using the words create and model. Do you use AMF to create an API model? Or do you use AMF to create an API? Or do you use AMF to model an API? We should clarify the verb and noun use of model

File: docs/other/model-in-memory.mdx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+In the following code example we're going to create a generic 'Music Service API' with the following functionality:
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:21]: The following code example creates a generic "Music Service API," with the following endpoints:

File: docs/other/model-in-memory.mdx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+- `/me` endpoint to retrieve user's profile
+- `/me/playlists` endpoint to retrieve user's playlists
+- `/albums` endpoint to retrieve a specific album or a list of albums by their IDs
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:25]: - `/me`: Retrieves a user's profile
- `/me/playlists`: Retrieves user's playlists
- `/albums` : Retrieves a specific album or a list of albums by their IDs

File: docs/other/model-in-memory.mdx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+As our focus is on building the API, all responses are going to be of type string, but you can further specify
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:27]: For the purposes of this API, all responses are returned as strings, but you can further specify requests and responses by defining a schema for an album, playlist, or user.

File: docs/other/payload-validation.mdx [hunk 9]
```diff
@@ -0,0 +1,124 @@
+---
+id: payload_validation
+title: Payload validation with AMF
+---
+
+AMF can be used to validate incoming payloads to an API already deployed and running.
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:6]: You can use AMF to validate incoming payloads to an API already deployed and running.

File: docs/other/payload-validation.mdx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+For example, consider a RAML API that declares:
+
+- a `User` schema with `firstname`, `lastname` and `age` properties
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:9]: `lastname`, and `age` properties

File: docs/other/payload-validation.mdx [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+      lastname:  string
+      age:       number
+/users:
+  post:
+    body:
+      application/json:
+        type: User
+```
+
+How can you confirm that a given payload is a valid `User`? You can do that with AMF payload validation.
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:29]: To confirm that a given payload is a valid `User`, you can use AMF payload validation. To do so:

File: docs/other/payload-validation.mdx [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+To do that, you first need to go through certain steps:
+
+1. Get the AMF payload shape
+2. Create a `PayloadValidator`
+3. Validate payloads
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:34]: 1. Get the AMF payload model.
2. Create a `PayloadValidator`.    or Create a payload model [or schema?]
3. Validate the payloads.

File: docs/other/payload-validation.mdx [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+
+## 1. Get the AMF payload model
+When AMF parses an API:
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:38]: When AMF parses an API, it:

File: docs/other/payload-validation.mdx [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+- Returns a `Document` that encodes a `WebApi`, which is the AMF model of the original API
+- Abstracts the HTTP methods to `Operation` objects
+
+![api to webapi model](../assets/images/other/api-to-webapi.png)
+
+The [AMF Model Documentation](https://github.com/aml-org/amf/blob/develop/documentation/model.md#webapi) has more information
+on each intermediate model, but at the end of the chain there is the **Payload shape** we want to use.
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:45]: The [AMF Model Documentation](https://github.com/aml-org/amf/blob/develop/documentation/model.md#webapi) has more information
on each intermediate model. 

note: I'm not sure what , but at the end of the chain there is the **Payload shape** we want to use... Is there some other way to phrase this?



Also, is it payload shape or payload model?

File: docs/other/payload-validation.mdx [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+How can you confirm that a given payload is a valid `User`? You can do that with AMF payload validation.
+To do that, you first need to go through certain steps:
+
+1. Get the AMF payload shape
+2. Create a `PayloadValidator`
+3. Validate payloads
+
+
+## 1. Get the AMF payload model
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:37]: ## Get the AMF payload model

I don't think we need--or have been using--numbers in the headings. (Please correct me if I'm wrong... 😜 )

File: docs/other/model-in-memory.mdx (hunk 1 above)
Comment (IsaacEldridge) [reply docs/other/model-in-memory.mdx:3]: Also @arielmirra I'm sorry I haven't noticed this until now, but I think we've been going back and forth on title and heading capitalization. I think we should standardize and use title case or sentence case for titles and headings. 

That means:

Creating a Model in Memory Or Creating a model in memory

and similarly throughout. If it's easier, I can create a separate ticket to PR that

File: docs/other/payload-validation.mdx [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+When AMF parses an API:
+- Returns a `Document` that encodes a `WebApi`, which is the AMF model of the original API
+- Abstracts the HTTP methods to `Operation` objects
+
+![api to webapi model](../assets/images/other/api-to-webapi.png)
+
+The [AMF Model Documentation](https://github.com/aml-org/amf/blob/develop/documentation/model.md#webapi) has more information
+on each intermediate model, but at the end of the chain there is the **Payload shape** we want to use.
+
+## 2. Create a PayloadValidator
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:47]: ## Create a payload validator

File: docs/other/payload-validation.mdx [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+  def parameterValidator(mediaType: String): ClientOption[PayloadValidator]
+
+  def parameterValidator(mediaType: String, exec: BaseExecutionEnvironment): ClientOption[PayloadValidator]
+
+  def parameterValidator(mediaType: String, env: Environment): ClientOption[PayloadValidator]
+}
+```
+
+This interface defines methods to create a `PayloadValidator` from an `AnyShape` given the `mediaType` and optionally the
+`BaseExecutionEnvironment` or `Environment` objects.
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:69]: This interface defines methods to create a `PayloadValidator` from an `AnyShape` class, given the `mediaType` and optional `BaseExecutionEnvironment` or `Environment` objects.

File: docs/other/payload-validation.mdx [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+- Returns a `Document` that encodes a `WebApi`, which is the AMF model of the original API
+- Abstracts the HTTP methods to `Operation` objects
+
+![api to webapi model](../assets/images/other/api-to-webapi.png)
+
+The [AMF Model Documentation](https://github.com/aml-org/amf/blob/develop/documentation/model.md#webapi) has more information
+on each intermediate model, but at the end of the chain there is the **Payload shape** we want to use.
+
+## 2. Create a PayloadValidator
+The Payload shape (or schema) is a specific implementation of the [AnyShape](https://github.com/aml-org/amf/blob/develop/documentation/model.md#anyshape) class,
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:48]: earlier this said payload model

File: docs/other/payload-validation.mdx [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+}
+```
+
+This interface defines methods to create a `PayloadValidator` from an `AnyShape` given the `mediaType` and optionally the
+`BaseExecutionEnvironment` or `Environment` objects.
+
+There are two distinct payload validators:
+
+- **`payloadValidator`**
+    - Validates types, values and constraints in strict mode
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:74]: types, values, and constraints

File: docs/other/payload-validation.mdx [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+- **`parameterValidator`**
+    - A best-effort validation, fails only if there is no applicable schema
+    - Used for validating query parameters, where different types may be sent as strings
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:77]: - Validates query parameters, where different types may be sent as strings; fails only if there is no applicable schema

File: docs/other/payload-validation.mdx [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+
+
+## 3. Validate payloads
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:80]: the payloads

File: docs/other/payload-validation.mdx [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+You can use the following methods of PayloadValidator to validate a payload:
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:81]: To validate a payload with `PayloadValidator`, use the following methods:

File: docs/other/payload-validation.mdx [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+
+- **`isValid(mediaType: String, payload: String): ClientFuture[Boolean]`**
+    - Validates a payload given its media type in fail-fast mode
+    - Stops execution when it encounters any violation
+    - Used for checking payload validity
+- **`validate(mediaType: String, payload: String): ClientFuture[ValidationReport]`**
+    - Validates a payload given its media type and returns a `ValidationReport` Object containing detailed information about all errors found
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:88]: returns a `ValidationReport` object containing detailed information about any errors

File: docs/other/payload-validation.mdx [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+- **`validate(payloadFragment: PayloadFragment): ClientFuture[ValidationReport]`**
+    - Validates a payload from an [AMF PayloadFragment](https://github.com/aml-org/amf/blob/develop/documentation/model.md#payloadfragment)
+- **`syncValidate(mediaType: String, payload: String): ValidationReport`**
+    - Validates a payload synchronously
+    - Returns the `ValidationReport` directly
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:93]: cut directly

File: docs/other/payload-validation.mdx [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+
+
+### Code examples of payload validation
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:96]: ## Example code for a payload validator

File: docs/other/payload-validation.mdx [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+The following code examples validate the previously mentioned `User` schema with the payload validation methods:
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:97]: The following code example validates the previously mentioned `User` schema with the payload validation methods:

**Review by IsaacEldridge — CHANGES_REQUESTED** [review]
@arielmirra back to you

File: docs/other/model-in-memory.mdx (hunk 1 above)
Comment (arielmirra) [reply docs/other/model-in-memory.mdx:3]: You're right @IsaacEldridge! if you can take a look into it in a different PR that would be nice, if you need me to create a ticket let me know 👍🏽

File: docs/other/model-in-memory.mdx (hunk 2 above)
Comment (arielmirra) [reply docs/other/model-in-memory.mdx:5]: mmmm not really, it can be ommited.

**Review by arielmirra — COMMENTED** [review]
@IsaacEldridge check the comments I've responded to

File: docs/other/payload-validation.mdx [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+
+## Validate the payloads
+To validate a payload with `PayloadValidator`, use the following methods:
+
+- **`isValid(mediaType: String, payload: String): ClientFuture[Boolean]`**
+    - Validates a payload given its media type in fail-fast mode
+    - Stops execution when it encounters any violation
+    - Used for checking payload validity
+- **`validate(mediaType: String, payload: String): ClientFuture[ValidationReport]`**
+    - returns a `ValidationReport` object containing detailed information about any errors
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:86]:
```suggestion
    - Returns a `ValidationReport` object containing detailed information about any errors
```

File: docs/other/model-in-memory.mdx [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+
+You can use an [AMF WebApi](https://github.com/aml-org/amf/blob/develop/documentation/model.md#webapi) in memory to:
+
+- Edit an already parsed API
+- Power a visual API designer
+- Create programmatic API tests
+
+
+### Create a WebApi with AMF
+The following code example creates a generic "Music Service API", with the following endpoints:
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:16]:
```suggestion
The following code example creates a generic "Music Service API" with the following endpoints:
```

File: docs/other/model-in-memory.mdx [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+---
+
+AMF allows you to parse existing APIs, but you can also use it to model fluent, descriptive APIs in memory.
+
+You can use an [AMF WebApi](https://github.com/aml-org/amf/blob/develop/documentation/model.md#webapi) in memory to:
+
+- Edit an already parsed API
+- Power a visual API designer
+- Create programmatic API tests
+
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:13]: extra space here

File: docs/other/model-in-memory.mdx [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+
+### Create a WebApi with AMF
+The following code example creates a generic "Music Service API", with the following endpoints:
+
+- `/me`: Retrieves a user's profile
+- `/me/playlists`: Retrieves user's playlists
```
Comment (IsaacEldridge) [inline docs/other/model-in-memory.mdx:19]:
```suggestion
- `/me/playlists`: Retrieves a user's playlists
```

File: docs/other/payload-validation.mdx [hunk 31]
```diff
@@ -0,0 +1,122 @@
+---
+id: payload_validation
+title: Payload validation with AMF
+---
+
+You can use AMF to validate incoming payloads to an API already deployed and running.
+For example, consider a RAML API that declares:
+
+- a `User` schema with `firstname`, `lastname`, and `age` properties
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:9]:
```suggestion
- A `User` schema with `firstname`, `lastname`, and `age` properties
```

File: docs/other/payload-validation.mdx [hunk 32]
```diff
@@ -0,0 +1,122 @@
...
+- a `/users` endpoint with an HTTP POST method that requires a `User` object as body parameter
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:10]:
```suggestion
- A `/users` endpoint with an HTTP POST method that requires a `User` object as a body parameter
```

File: docs/other/payload-validation.mdx [hunk 33]
```diff
@@ -0,0 +1,122 @@
...
+
+## Get the AMF payload shape
+When AMF parses an API, it:
+- Returns a `Document` that encodes a `WebApi`, which is the AMF WebApi graph model of the original API
+- Abstracts the HTTP methods to `Operation` objects
+
+![api to webapi model](../assets/images/other/api-to-webapi.png)
+
+The [AMF Model Documentation](https://github.com/aml-org/amf/blob/develop/documentation/model.md#webapi) has more information
+on each intermediate model. The Payload shape we'll use can be obtained from an `Operation` `Request` or `Response` object.
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:44]:
```suggestion
on each intermediate model. The payload shape we'll use can be obtained from an `Operation` `Request` or `Response` object.
```

File: docs/other/payload-validation.mdx [hunk 34]
```diff
@@ -0,0 +1,122 @@
...
+
+## Create a `PayloadValidator`
+The Payload shape (or schema) is a specific implementation of the [AnyShape](https://github.com/aml-org/amf/blob/develop/documentation/model.md#anyshape) class,
```
Comment (IsaacEldridge) [inline docs/other/payload-validation.mdx:47]:
```suggestion
The payload shape (or schema) is a specific implementation of the [AnyShape](https://github.com/aml-org/amf/blob/develop/documentation/model.md#anyshape) class,
```

**Review by IsaacEldridge — CHANGES_REQUESTED** [review]
Just a few minor changes and then good to merge

**Review by IsaacEldridge — APPROVED** [review]

**Pull request merged**